LEARNER-3718: Refactor marketing pipelines.

robrap opened the pull request:
This is a draft of refactoring the marketing pipeline code.

I'll post my own questions soon when I am available.  Some of those are in TODO comments that I added as I went.

macdiesel commented on edxpipelines/patterns/marketing_stages.py:
```diff
@@ -0,0 +1,254 @@
+        create_tag: True if a tag should first be created, False otherwise.
+
+    Returns:
+        The new stage.
```
Should be: gomatic.Stage

robrap replied:
Ok.

macdiesel commented on edxpipelines/patterns/marketing_stages.py:
```diff
@@ -0,0 +1,254 @@
+    Creates a stage that creates and pushes source to Acquia.
+
+    Arguments:
+        pipeline: The pipeline to which the stage is being added.
```
Please include the types on the arguments. For pipeline it should be gomatic.Pipeline

robrap replied:
Will do.

robrap commented on config.yml:
```diff
@@ -110,7 +110,7 @@ tools:
   ## start Marketing Pipelines
   # Deploying to Stage and Production
-  - script: edxpipelines/pipelines/deploy_marketing_site.py
+  - script: edxpipelines/pipelines/deploy_marketing_site_production.py
```
Thanks @macdiesel.  I can definitely fix.  I have some questions around the approach.

The current deploy_marketing_site_production.py pipeline first deploys to a stage environment (test) and then deploys to prod.  The deploy_marketing_site_development.py deploys to a single environment.  I could make a single script that could do either.

I would add a new prod config (like the this [config for acceptance](https://github.com/edx/edx-gomatic/blob/master/edxpipelines/pipelines/config/edx-mktg/drupal-acceptance-environment.yml)), only with the following contents:
```
DRUPAL_ENVIRONMENT = 'prod'
DRUPAL_STAGE_ENVIRONMENT = 'test'
```
I would use the existence of `DRUPAL_STAGE_ENVIRONMENT` to determine whether or not it should do a two environment deployment or a single deployment.

Additionally, there are some other differences between our current two environment deployments and single environment deployments:
- We backup the database during our two environment deployments, but not for a single environment deployment.
- We deploy from tags for the two environment deployment, and not for a single environment deployment.
For now, I think I would hardcode these rules based off the existence of `DRUPAL_STAGE_ENVIRONMENT` to keep things simpler and have less likelihood of config related issues.  In the future these could be added as config options if they need to be adjusted any more than this.  

Does this all make sense?

robrap replied:
I enhanced [the comment for this pipeline](https://github.com/edx/edx-gomatic/blob/bb42c08d7ecf52db4cfbbfc2373a96b29f60e984/edxpipelines/pipelines/deploy_marketing_site_production.py#L24-L28) and made it configurable even though it is only used for stage/prod at this time.  I opted to leave it called "production" because there isn't a great name for this that wouldn't be more confusing at this time.

macdiesel replied:
@robrap This was probably more of a nit than anything else. I'm open to whatever approach you guys need to use and to ensure you guys were aware that this was a bit outside of the paradigm we typically use.

robrap commented on edxpipelines/patterns/marketing_jobs.py:
```diff
@@ -0,0 +1,47 @@
+from edxpipelines.patterns import tasks
+
+
+def generate_job_fetch_tag_name(stage, drupal_env):
```
Can do.

robrap commented on edxpipelines/patterns/marketing_jobs.py:
```diff
@@ -0,0 +1,47 @@
+            be fetched.
+
+    Returns:
+        The new job.
```
Will do.

robrap commented on edxpipelines/patterns/marketing_jobs.py:
```diff
@@ -0,0 +1,47 @@
+    fetch_tag_job.ensure_artifacts(
+        set([BuildArtifact('target/{tag_name}.txt'.format(
+            tag_name=constants.ACQUIA_ENV_TAG_NAME.format(drupal_env=drupal_env)
+        ))])
```
Ok.

robrap commented on edxpipelines/patterns/marketing_stages.py:
```diff
@@ -0,0 +1,254 @@
+        tasks.generate_target_directory(deploy_job_for_drupal_env)
+
+        # fetch the tag name
+        # TODO: Temporary Comment: This call to
```
@Dillon-Dumesnil: FYI: I plan to delete this comment, but if I finished it, it meant to highlight the following...  In the original script, the following was only done once for both stage (test) and prod in the production pipeline.  It seems better to keep the code clean and not try to share this across those jobs.

robrap commented on edxpipelines/pipelines/deploy_marketing_site_development.py:
```diff
@@ -9,83 +10,30 @@
-            'MARKETING_REPOSITORY_VERSION': config['mktg_repository_version'],
-        }
-    )
+    # TODO: Need to determine if it is ok to keep the pipeline stages at this
```
@macdiesel: Based on earlier comment, I think I am going to move/share the pipeline code for both two environment and single environment deployments.

robrap commented on edxpipelines/pipelines/deploy_marketing_site_production.py:
```diff
@@ -0,0 +1,42 @@
+    """
+    Install pipelines that can deploy the edx-mktg site.
+    """
+    # TODO: Determine if we can rename the pipeline from xxx to xxx-prod.
```
@macdiesel: Any concern with having a different name for the pipeline before and after?

robrap commented on edxpipelines/constants.py:
```diff
@@ -156,27 +156,16 @@
 FETCH_TAG_JOB_NAME = 'fetch_current_tag_names_job'
 PUSH_TO_ACQUIA_STAGE_NAME = 'push_to_acquia'
```
@Dillon-Dumesnil @macdiesel: I'm unclear why we decided to include `{drupal_env}` in some, but not all of the marketing stage/job names.  Some additional possibilities incude:
- `'fetch_current_tag_names_{drupal_env}_job'`
- `'push_to_acquia_for_{drupal_env}'`
- `'rollback_{drupal_env}_deployment'`
What are the pros/cons of a shared stage name?

FYI: On a related note, @Dillon-Dumesnil and broken tests made it clear I need to update the rollback scripts as well.

robrap commented:
@macdiesel @Dillon-Dumesnil: This is ready for review again.
1. I added refactoring of rollback script (was scripts) to fix test and complete refactoring.
2. I made the "production" pipeline more configurable and dropped that hard-coded constants for PROD and STAGE.
3. I fixed code review comments and tests.

Once we are all good with this, I'll squash and then work on merge with Dillon's branch and latest fixes.  Trying to get this in asap, so depending on your availability @macdiesel, I may have to squash early and merge and we'll complete the review on @Dillon-Dumesnil's branch.

robrap commented:
@macdiesel: Thanks.
@Dillon-Dumesnil: Rebase is complete.  This is ready for final review and merge.

Dillon-Dumesnil commented on edxpipelines/patterns/marketing_stages.py:
```diff
@@ -0,0 +1,299 @@
+    push_to_acquia_job.add_task(
+        tasks.bash_task(
+            """\
+            /usr/bin/git config user.email "admin@edx.org" &&
```
We are running these config commands twice if create_tag is True. I'm honestly not even sure if these are necessary with the way the github key works now. @macdiesel Can you speak to if we should still need to do this config?

robrap replied:
We are leaving this for now.

Dillon-Dumesnil commented on edxpipelines/patterns/marketing_stages.py:
```diff
@@ -0,0 +1,299 @@
+    marketing_tasks.generate_drupal_deploy(
+        rollback_job,
+        drupal_env,
+        '$(cat ../{artifact_path}/{tag_name}.txt)'.format(
```
Should this follow the same format of defining it as a deploy_source variable? It might help with readability

robrap replied:
This is written for tags only for now.

Dillon-Dumesnil commented on edxpipelines/pipelines/deploy_marketing_site_production.py:
```diff
@@ -0,0 +1,52 @@
+    marketing_stages.generate_push_to_acquia(pipeline, create_tag=True)
+
+    # Stages to deploy to stage environment
+    marketing_stages.generate_stages_deploy_marketing(pipeline, stage_env, backup_db=True, use_tag=True)
```
**IMPORTANT** Are we pausing the pipeline after this stage? Currently, the pipeline pauses after it finishes deploying to stage so people can review their work before going to production.

robrap replied:
Added in manual approval.

robrap commented on edxpipelines/constants.py:
```diff
@@ -151,34 +151,22 @@
-CLEAR_PROD_CACHES_JOB_NAME = 'clear_prod_caches_job'
-DEPLOY_PROD_STAGE_NAME = 'deploy_to_prod'
-DEPLOY_PROD_JOB_NAME = 'deploy_to_prod_job'
+BACKUP_ACQUIA_ENV_DATABASE_STAGE_NAME = 'backup_{drupal_env}_database'
```
@Dillon-Dumesnil: As you continue your work on this, I found it confusing that we use `Acquia`, `Drupal`, and `Site` often to mean the same thing (e.g. `acquia_env`, `drupal_env`, and sometimes `site_env`).  If you get a chance in your future work, it would be great to start with a comment that details when each term should be used (if at all), so we have a target to move toward.  Then, you could refactor as much or as little as you want toward that target.

Dillon-Dumesnil replied:
I believe they all refer to the same thing and would be happy to refactor it. Do you have a preference on acquia_env vs drupal_env @robrap ? I don't think site_env is as informative.

robrap replied:
I'd go with `acquia_env` given that `@acquiaenvs` is used in HipChat.
